File: src/main/java/com/cxxcxx/zinecraft/core/entity/ExampleBlockEntity.java
package com.cxxcxx.zinecraft.core.entity;

import net.minecraft.core.BlockPos;
import net.minecraft.core.HolderLookup.Provider;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.state.BlockState;
import org.jetbrains.annotations.NotNull;

public final class ExampleBlockEntity extends BlockEntity {
  private int clicks;

  public ExampleBlockEntity(@NotNull BlockPos blockPos, @NotNull BlockState blockState) {
    super(ModBlockEntity.EXAMPLE_BLOCK_ENTITY.get(), blockPos, blockState);
  }

  public final int getClicks() {
    return this.clicks;
  }

  public final void incrementClicks() {
    int i = this.clicks++;
    this.setChanged();
  }

  protected void loadAdditional(@NotNull CompoundTag nbt, @NotNull Provider registryLookup) {
    super.loadAdditional(nbt, registryLookup);
    this.clicks = nbt.getInt("clicks");
  }

  protected void saveAdditional(@NotNull CompoundTag nbt, @NotNull Provider registryLookup) {
    nbt.putInt("clicks", this.clicks);
    super.saveAdditional(nbt, registryLookup);
  }
}

File: src/main/kotlin/com/cxxcxx/zinecraft/core/entity/ExampleBlockEntity.kt
package com.cxxcxx.zinecraft.core.entity

import net.minecraft.core.BlockPos
import net.minecraft.core.HolderLookup
import net.minecraft.nbt.CompoundTag
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.state.BlockState


class ExampleBlockEntity(
  blockPos: BlockPos,
  blockState: BlockState,
) : BlockEntity(ModBlockEntity.EXAMPLE_BLOCK_ENTITY, blockPos, blockState) {

  private var clicks = 0

  fun getClicks(): Int {
    return clicks
  }

  fun incrementClicks() {
    clicks++
    setChanged()
  }

  override fun saveAdditional(nbt: CompoundTag, registryLookup: HolderLookup.Provider) {
    nbt.putInt("clicks", clicks)

    super.saveAdditional(nbt, registryLookup)
  }

  override fun loadAdditional(nbt: CompoundTag, registryLookup: HolderLookup.Provider) {
    super.loadAdditional(nbt, registryLookup)

    clicks = nbt.getInt("clicks")
  }
}